File: 백준/Gold/2239. 스도쿠/스도쿠.py
# 행검사
def row_check(r, num):
    for x in range(9):
        if num == sdoku[r][x]:
            return False
    return True


def col_check(c, num):
    # 열검사
    for x in range(9):
        if num == sdoku[x][c]:
            return False
    return True


def three_check(r, c, num):
    # 3x3 검사
    nc = (c // 3) * 3
    nr = (r // 3) * 3
    for x in range(3):
        for y in range(3):
            if sdoku[nr + x][nc + y] == num:
                return False
    return True


def dfs(depth):
    if depth >= len(zero_p):
        for k in range(9):
            print(''.join(map(str, sdoku[k])))
        exit()

    nr, nc = zero_p[depth]  # 리스트에서 하나씩 튜플로 꺼낸다.
    for j in range(1, 9 + 1):
    	#dfs를 돌기전에 검사하여 조건에 맞는 것만 dfs를 돈다
        if row_check(nr, j) and col_check(nc, j) and three_check(nr, nc, j):
            sdoku[nr][nc] = j
            dfs(depth + 1)
            sdoku[nr][nc] = 0


sdoku = []
zero_p = []  # 2차원 9*9 리스트를 1차원적으로 생각하여 dfs를 돌기 때문에 좌표들을 튜플형식으로 넣는다.
for i in range(9):
    temp = list(map(int, input()))
    for j in range(len(temp)):
        if temp[j] == 0:
            zero_p.append((i, j))  # 0의 좌표들을 저장한다.(1차원 리스트에 튜플로 저장)
    sdoku.append(temp)

dfs(0)

File: 백준/Gold/2239. 스도쿠/스도쿠.java
import java.util.*;
import java.io.*;

class Main {
	
	public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        
        int[][] board = new int[9][9];
        List<int[]> zeros = new ArrayList<>(); 

        
        for (int i = 0; i < 9; i++) {
        	char[] input = br.readLine().toCharArray();
        	
        	for (int j = 0; j < 9; j++) {
        		board[i][j] = input[j] - '0';
        		
        		if (board[i][j] == 0) zeros.add(new int[] {i, j});
        	}
        }
        
        dfs(board, 0, zeros);
	}
	
	private static void dfs(int[][] board, int depth, List<int[]> zeros) {
		if (depth == zeros.size()) {
			for (int i = 0; i < 9; i++) {
			    for (int j = 0; j < 9; j++) {
			    	System.out.print(board[i][j]);
			    }
			    System.out.println();
			}
			
			System.exit(0);
		}
		
		int x = zeros.get(depth)[0];
		int y = zeros.get(depth)[1];
		
		for (int i = 1; i <= 9; i++) {
			if (checkRow(board, i, x) && checkCol(board, i, y) && checkSquare(board, i, x, y)) {
				board[x][y] = i;
				dfs(board, depth + 1, zeros);
				board[x][y] = 0;
			}
		}
	}
	
	private static boolean checkRow(int[][] board, int num, int r) {
		for (int i = 0; i < 9; i++) {
			if (board[r][i] == num) return false;
		}
		
		return true;
	}
	
	private static boolean checkCol(int[][] board, int num, int c) {
		for (int i = 0; i < 9; i++) {
			if (board[i][c] == num) return false;
		}
		
		return true;
	}
	
	private static boolean checkSquare(int[][] board, int num, int x, int y) {
		int startX = x / 3 * 3;
		int startY = y / 3 * 3;
		
		for (int i = startX; i < startX + 3; i++) {
			for (int j = startY; j < startY + 3; j++) {
				if (board[i][j] == num) return false;
			}
		}
		
		return true;
	}
}
